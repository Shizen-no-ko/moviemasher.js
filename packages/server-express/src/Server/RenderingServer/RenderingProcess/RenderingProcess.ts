import { ApiCallback, RenderingOptions, RenderingResult } from "@moviemasher/moviemasher.js"

export interface RenderingProcessInput extends RenderingOptions {
  cacheDirectory: string
  renderingDirectory: string
  fileDirectory: string
}

export interface RenderingCallback {
  callback?: ApiCallback
  id?: string
}

export interface RunResult {
  results: RenderingResult[]
}

export interface RenderingProcessOptions extends RenderingProcessInput, RenderingCallback {}

export interface RenderingProcessArgs extends Required<RenderingProcessInput>, RenderingCallback {}

export interface RenderingProcess {
  runPromise: ()=> Promise<RunResult>
}
