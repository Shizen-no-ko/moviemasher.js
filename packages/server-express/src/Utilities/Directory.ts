import fs from 'fs'

const directoryLatest = (directory: string, extension: string) => {
  if (!fs.existsSync(directory)) return

  const files = fs.readdirSync(directory)
  const filesWithExt = files.filter(file => file.endsWith(extension)).sort()
  const count = filesWithExt.length
  const file = count && filesWithExt[count - 1]
  if (!file) return

  return `${directory}/${file}`
}

const Directory = {
 latest: directoryLatest
}

export { Directory, directoryLatest }
