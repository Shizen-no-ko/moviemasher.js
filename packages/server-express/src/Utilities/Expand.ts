import fs from 'fs'
import path from 'path'
import { execSync } from 'child_process'
import { ExtJson, JsonObject, ExtText } from '@moviemasher/moviemasher.js'

export const expandCommand = (command?: string): string => {
  if (!command) return ''

  if (command.endsWith(ExtText)) return expandFile(command)

  if (command.startsWith('/')) return execSync(command).toString()
  return command
}

export const expandFile = (file?: string): string => {
  return file ? fs.readFileSync(file).toString() : ''
}

export const expandToString = (string?: string): string => {
  if (!string) return ''

  switch (string[0]) {
    case '.': { // relative path
      return expandToString(path.resolve(string))
    }
    case '/': { // absolute path
      return expandToString(expandFile(string))
    }
  }
  // a command that returns a json string
  return expandToString(expandCommand(string))
}

export const expandPath = (string: string): string => {
  return string.startsWith('.') ? path.resolve(string) : string
}

export const expandToJson = (config: string): JsonObject => {
  if (!config) return {}

  if (config.endsWith(ExtJson)) return expandToJson(expandFile(config))

  switch (config[0]) {
    case '.':
    case '/': { // path to script, since it doesn't end in .json
      return expandToJson(expandCommand(expandPath(config)))
    }
    case '{': { // json string
      return JSON.parse(config)
    }
  }
  // failed to expand to JSON
  return {}
}
