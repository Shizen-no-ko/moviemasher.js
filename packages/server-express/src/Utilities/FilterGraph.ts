import {
  GraphFilter, FilterGraph, AVType, FilterChain, OutputOptions, GraphType, ValueObject
} from "@moviemasher/moviemasher.js"

import { CommandArgs, CommandInput } from "../Command/Command"

const filterGraphToCommandArgs = (filterGraph: FilterGraph, options: OutputOptions, destination: string, graphType = GraphType.Cast): CommandArgs => {
  const { filterChains, duration, avType } = filterGraph
  const inputs: CommandInput[] = []
  const complexFilter: GraphFilter[] = []
  if (!duration) throw 'duration'

  const outputOptions: ValueObject = options.options || {}
  if (graphType !== GraphType.Cast) outputOptions.t ||= duration
  const output: OutputOptions = { format: options.format, options: outputOptions }
  if (!avType || avType === AVType.Audio) {
    if (options.audioCodec) output.audioCodec = options.audioCodec
    if (options.audioBitrate) output.audioBitrate = options.audioBitrate
    if (options.audioChannels) output.audioChannels = options.audioChannels
    if (options.audioRate) output.audioRate= options.audioRate
  }
  if (!avType || avType === AVType.Video) {
    if (options.videoCodec) output.videoCodec = options.videoCodec
    if (options.videoRate) output.videoRate = options.videoRate
  }

  filterChains.forEach((filterChain: FilterChain) => {
    const { merger, filters, inputs, files } = filterChain
    files.forEach(graphFile => {
      const { type, file } = graphFile

    })
    inputs.push(...inputs)
    complexFilter.push(...filters)
    if (merger) complexFilter.push(merger)
  })
  const commandOptions: CommandArgs = {
    inputs, complexFilter, output, destination
  }
  return commandOptions
}

export { filterGraphToCommandArgs }
