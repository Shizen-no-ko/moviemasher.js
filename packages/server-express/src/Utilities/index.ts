export * from "./Directory"
export * from "./NodePreloader"
export * from "./Rendering"
