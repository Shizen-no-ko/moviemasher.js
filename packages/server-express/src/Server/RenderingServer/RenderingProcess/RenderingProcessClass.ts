import fs from 'fs'
import path from 'path'

import {
  fetchCallback,
  idGenerate, Mash, MashFactory, OutputFactory,
  RenderingOutputArgs,
  Errors,
  RenderingOutput,
  outputDefaultPopulate,
  RenderingCommandOutput, RenderingResult, OutputType, CommandOutput, OutputFormat, outputDefaultVideo,

} from "@moviemasher/moviemasher.js"
import { RunningCommandFactory } from "../../../RunningCommand/RunningCommandFactory"
import { RenderingProcess, RenderingProcessArgs, RunResult } from "./RenderingProcess"
import { NodePreloader } from '../../../Utilities/NodePreloader'
import { CommandResult } from '../../../RunningCommand/RunningCommand'

const outputDefaultOverDefaults = (overrides?: CommandOutput): RenderingCommandOutput => {
  const object = overrides || {}
  const commandOutput = outputDefaultVideo(overrides)
    return {
      ...commandOutput,
      ...object,
    }
}

const outputDefaultAudioConcat = (overrides?: CommandOutput): RenderingCommandOutput => {
  const object = overrides || {}
  const baseOptions: CommandOutput = { ...object, format: OutputFormat.AudioConcat }
  return outputDefaultOverDefaults(baseOptions)
}

const outputDefaultVideoConcat = (overrides?: CommandOutput): RenderingCommandOutput => {
  const object = overrides || {}
  const baseOptions: CommandOutput = { ...object, format: OutputFormat.VideoConcat }
  return outputDefaultOverDefaults(baseOptions)
}


class RenderingProcessClass implements RenderingProcess {
  constructor(args: RenderingProcessArgs) {
    this.args = args
  }

  args: RenderingProcessArgs

  directoryPromise(): Promise<void> {
    const { idDirectory } = this
    // console.log(this.constructor.name, "directoryPromise", idDirectory)
    const definitionsJson = JSON.stringify(this.args.definitions)
    const mashJson = JSON.stringify(this.args.mash)
    return fs.promises.mkdir(idDirectory, { recursive: true }).then(() => {
      const mashPath = path.join(idDirectory, 'mash.json')
      return fs.promises.writeFile(mashPath, mashJson).then(() => {
        const definitionsPath = path.join(idDirectory, 'definitions.json')
        return fs.promises.writeFile(definitionsPath, definitionsJson)
      })
    })
  }

  private _id?: string
  get id(): string {
    if (this._id) return this._id

    return this._id = this.args.id || idGenerate()
  }

  get idDirectory(): string {
    const { renderingDirectory } = this.args
    return path.join(renderingDirectory, this.id)
  }

  private _mashInstance?: Mash
  get mashInstance(): Mash {
    if (this._mashInstance) return this._mashInstance

    const { definitions, mash } = this.args
    return this._mashInstance = MashFactory.instance(mash, definitions)
  }

  outputInstance(commandOutput: RenderingCommandOutput): RenderingOutput {
    const { outputType } = commandOutput
    const { cacheDirectory, fileDirectory } = this.args
    const preloader = new NodePreloader(cacheDirectory, fileDirectory)
    const { mashInstance } = this
    mashInstance.preloader = preloader
    const args: RenderingOutputArgs = {
      commandOutput, cacheDirectory, mash: mashInstance,
    }
    return OutputFactory[outputType](args)
  }


  outputTypeDestination(renderingOutput: RenderingOutput, commandOutput: RenderingCommandOutput): string {
    const { outputType  } = renderingOutput
    const { format, extension } = commandOutput
    switch (outputType) {
      case OutputType.VideoSequence:
        const fps = Number(commandOutput.videoRate)
        const {duration} = renderingOutput
        const framesMax = Math.floor(fps * duration) - 2
        const begin = 1

        const lastFrame = begin + (framesMax - begin)
        const padding = String(lastFrame).length
        return `${this.idDirectory}/%0${padding}d.${extension || format}`

      default: return `${this.idDirectory}/${outputType}.${extension || format}`
    }
  }

  runPromise(): Promise<RunResult> {
    const results: RenderingResult[] = []
    const runResult: RunResult = { results }

    let promise = this.directoryPromise()
    const { outputs } = this.args
    outputs.forEach(renderingCommandOutput => {
      const commandOutput = outputDefaultPopulate(renderingCommandOutput)
      // console.log(this.constructor.name, "runPromise", renderingCommandOutput, commandOutput)
      const { outputType } = commandOutput
      const output = this.outputInstance(commandOutput)
      const outputPromise = output.commandArgsPromise(results).then(commandArgs => {
        // console.log(this.constructor.name, "runPromise filterGraphs", filterGraphs)
        if (commandArgs.length > 1) {
          commandArgs.forEach(commandArg => {
            console.log(this.constructor.name, "runPromise MULTIPLE commandArg", commandArg)
            throw Errors.unimplemented + 'multiple commands...'
          })
        } else {
          const [commandArg] = commandArgs
          // console.log(this.constructor.name, "runPromise commandArg", commandArg)
          // return Promise.resolve()

          // const { format, extension } = commandOutput
          // `${this.idDirectory}/${outputType}.${extension || format}`
          const destination = this.outputTypeDestination(output, commandOutput)

          const command = RunningCommandFactory.instance(this.id, commandArg)

          command.addListener('error', error => {
            // console.log(error, commandArg, command.command._getArguments(), destination)
            throw String(error)
          })

          return command.runPromise(destination).then((result: CommandResult) => {
            const { error } = result
            // console.log(this.constructor.name, "runPromise runPromise then", result)
            if (error) throw error

            const renderingResult: RenderingResult = {
              ...result, destination, outputType
            }
            results.push(renderingResult)
            // return renderingResult
          })
        }
      })
      promise = promise.then(() => outputPromise)
    })

    const { callback } = this.args
    if (callback) promise = promise.then(() => fetchCallback(callback))
    return promise.then(() => runResult)
  }
}

export { RenderingProcessClass }
