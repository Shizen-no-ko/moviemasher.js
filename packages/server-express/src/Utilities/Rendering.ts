
import {
  DefinitionObject, DefinitionObjects,
  MashObject, TrackObject, ClipObject,
  DefinitionType, LoadType, TrackType, ValueObject, CommandOutputs, OutputType,
  RenderingInput, RenderingOutput, RenderingCommandOutput, NumberObject, outputDefaultPopulate,
} from "@moviemasher/moviemasher.js"
import { renderingProcessTestArgs } from "../../../../dev/test/Utilities/renderingProcessArgs"
import { RenderingProcessArgs } from "../Server/RenderingServer/RenderingProcess/RenderingProcess"

const uuid = require('uuid').v4


const renderingInput = (definition: DefinitionObject, clipObject: ValueObject = {}): RenderingInput => {
  const { type, id } = definition
  const definitionObject = {
    ...definition,
    type: type === DefinitionType.VideoSequence ? DefinitionType.Video : type
  }
  const definitions: DefinitionObjects = [definitionObject]
  const clip: ClipObject = renderingClipFromDefinition(definitionObject, clipObject)
  const track: TrackObject = {
    trackType: String(type) === String(DefinitionType.Audio) ? TrackType.Audio : TrackType.Video,
    clips: [clip]
  }
  const tracks: TrackObject[] = [track]
  const mashObject = { id }
  const mash: MashObject = { ...mashObject, tracks }
  return { mash, definitions }
}

const renderingDefinitionTypeCommandOutputs = (definitionType: DefinitionType) => {
  const outputs: CommandOutputs = []

  // TODO: support waveform generation
  // TODO: support font uploading
  switch (definitionType) {
    case DefinitionType.Audio: {
      outputs.push({ outputType: OutputType.Audio })
      // outputs.push({ outputType: OutputType.Waveform })
      break
    }
    case DefinitionType.Image: {
      outputs.push({ outputType: OutputType.Image })
      break
    }
    case DefinitionType.VideoSequence: {
      outputs.push({ outputType: OutputType.Audio })
      outputs.push({ outputType: OutputType.Image })
      outputs.push({ outputType: OutputType.ImageSequence })
      // outputs.push({ outputType: OutputType.Waveform })
      break
    }
    case DefinitionType.Font: {
      // outputs.push({ outputType: OutputType.Font })
      break
    }
  }
  return outputs
}

const renderingCommandOutputs = (commandOutputs: CommandOutputs): CommandOutputs => {
  const counts: NumberObject = {}
  return commandOutputs.map(output => {
    const { outputType } = output
    const populated = outputDefaultPopulate(output)
    if (!counts[outputType]) {
      counts[outputType] = 1
    } else {
      populated.basename ||= `${outputType}-${counts[outputType]}`
      counts[outputType]++
    }
    return populated
  })
}

const renderingOutputFile = (commandOutput: RenderingCommandOutput, extension?: string): string => {
  const { basename, format, extension: outputExtension, outputType } = commandOutput
  const ext = extension || outputExtension || format
  return `${basename || outputType}.${ext}`
}

const renderingSource = (commandOutput?: RenderingCommandOutput): string => {
  if (!commandOutput) {
    // console.log("renderingSource with no commandOutput")
    return ''
  }

  const { format, extension, outputType } = commandOutput
  const ext = extension || format
  if (outputType === OutputType.ImageSequence) return ''

  return `${outputType}.${ext}`
}

export const renderingProcessArgs = (id?: string): RenderingProcessArgs => {
  return {
    ...renderingProcessTestArgs(id),
    definitions: [], mash: {}, outputs: []
  }
}

export const renderingInputFromRaw = (loadType: LoadType, source: string, clip: ValueObject = {}): RenderingInput => {
  const definition = renderingDefinitionObject(loadType, source, String(clip.id), String(clip.label))
  return renderingInput(definition, clip)
}


export const renderingClipFromDefinition = (definition: DefinitionObject, overrides: ValueObject = {}): ClipObject => {
  const { id, type } = definition
  const clip: ClipObject = { definitionId: id || type, ...overrides }
  return clip
}

export const renderingDefinitionObject = (loadType: LoadType, source: string, definitionId?: string, label?: string): DefinitionObject => {
  const type: DefinitionType = definitionTypeFromRaw(loadType)
  const id = definitionId || uuid()
  const definition: DefinitionObject = { id, type, source, label }
  return definition
}



export const definitionTypeFromRaw = (loadType: LoadType): DefinitionType => {
  switch (loadType) {
    case LoadType.Audio: return DefinitionType.Audio
    case LoadType.Video: return DefinitionType.VideoSequence
    case LoadType.Image: return DefinitionType.Image
    case LoadType.Font: return DefinitionType.Font
  }
}


export {
  renderingOutputFile,
  renderingSource,
  renderingDefinitionTypeCommandOutputs,
  renderingCommandOutputs,
  renderingInput,
}
