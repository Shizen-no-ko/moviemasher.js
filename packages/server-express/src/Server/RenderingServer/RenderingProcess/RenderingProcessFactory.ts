import { RenderingProcessArgs } from "./RenderingProcess"
import { RenderingProcessClass } from "./RenderingProcessClass"

const renderingProcessInstance = (options: RenderingProcessArgs) => {
  return new RenderingProcessClass(options)
}

export { renderingProcessInstance }
