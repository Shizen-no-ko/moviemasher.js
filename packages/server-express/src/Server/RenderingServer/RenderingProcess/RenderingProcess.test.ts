import fs from 'fs'
import {
  RenderingCommandOutput, OutputType, LoadType, RenderingInput, OutputFormat,
  ExtPng, CommandOutputs, outputDefaultPopulate
} from "@moviemasher/moviemasher.js"

import { RenderingProcessArgs } from "./RenderingProcess"
import { renderingProcessInstance } from "./RenderingProcessFactory"
import { expectOutputFile } from "../../../../../../dev/test/Utilities/expectOutputFile"
import { renderingProcessTestArgs
} from "../../../../../../dev/test/Utilities/renderingProcessArgs"
import { TestRenderingsDir } from '../../../../../../dev/test/Setup/Constants'
import { renderingInputFromRaw } from '../../../Utilities/Rendering'
import { expandFile } from '../../../Utilities/Expand'

export function renderingProcessArgs(id?: string): RenderingProcessArgs {
  return {
    ...renderingProcessTestArgs(id),
    definitions: [], mash: {}, outputs: []
  }
}

describe("RenderingProcess", () => {
  const imageUrl = '../shared/image/globe.jpg'
  const videoUrl = '../shared/video.mp4'

  const baseOutput: RenderingCommandOutput = {
    outputType: OutputType.Image,
    format: OutputFormat.Png,
    extension: ExtPng,
  }
  describe("renders", () => {
    const renderingFiles = fs.readdirSync(TestRenderingsDir).filter(p => p.endsWith('.json'))
    test.each(renderingFiles)('%s', async (fileName) => {
      const json = expandFile(`${TestRenderingsDir}/${fileName}`)
      const args = JSON.parse(json) as RenderingProcessArgs
      const { id, outputs } = args
      const renderingArgs = { ...args, ...renderingProcessTestArgs(id) }
      const renderingProcess = renderingProcessInstance(renderingArgs)
      const runPromise = renderingProcess.runPromise()
      const checkPromise = runPromise.then(runResult => {
        const { results } = runResult
        const promises = outputs.map((output, index) => {
          const result = results[index]
          expect(result).toBeDefined()
          const { destination, error } = result
          expect(error).toBeFalsy()
          expect(destination).toBeDefined()

          return expectOutputFile(destination)
        })
        return Promise.all(promises)
      })
      await checkPromise
    })
  })
  describe("render", () => {
    test.skip("renders audio from trimmed video", async () => {
      const output = outputDefaultPopulate({ outputType: OutputType.Audio })
      const outputs: CommandOutputs = [output]
      const renderingInput: RenderingInput = renderingInputFromRaw(LoadType.Video, videoUrl, { trim: 10 })
      const processArgs: RenderingProcessArgs = {
        ...renderingProcessArgs('audio-from-trimmed-video'), outputs, ...renderingInput,
      }
      const renderingProcess = renderingProcessInstance(processArgs)
      const result = await renderingProcess.runPromise()

      await expectOutputFile(result.results[0].destination)
    })
  })

  describe("runPromise", () => {

    test("renders image from image with cover", async () => {
      const output = outputDefaultPopulate({ ...baseOutput, cover: true })
      const outputs: CommandOutputs = [output]
      const renderingInput: RenderingInput = renderingInputFromRaw(LoadType.Image, imageUrl)
      const processArgs: RenderingProcessArgs = {
        ...renderingProcessArgs('image-from-image-square-with-cover'), outputs, ...renderingInput,
      }
      // console.log(processArgs)
      const renderingProcess = renderingProcessInstance(processArgs)
      const result = await renderingProcess.runPromise()
      await expectOutputFile(result.results[0].destination)
    })


    test("renders video from image", async () => {
      const output = outputDefaultPopulate({ outputType: OutputType.Video, cover: false })
      const outputs: CommandOutputs = [output]
      const renderingInput: RenderingInput = renderingInputFromRaw(LoadType.Image, imageUrl)
      const processArgs: RenderingProcessArgs = {
        ...renderingProcessArgs('video-from-image-square'), outputs, ...renderingInput,
      }
      const process = renderingProcessInstance(processArgs)
      const result = await process.runPromise()
      await expectOutputFile(result.results[0].destination)
    })


    test.skip("renders image from trimmed video", async () => {
      const output = outputDefaultPopulate({ ...baseOutput })
      const outputs: CommandOutputs = [output]
      const renderingInput: RenderingInput = renderingInputFromRaw(LoadType.Video, videoUrl, { trim: 10 })
      const processArgs: RenderingProcessArgs = {
        ...renderingProcessArgs('image-from-trimmed-video'), outputs, ...renderingInput,
      }
      const renderingProcess = renderingProcessInstance(processArgs)
      const result = await renderingProcess.runPromise()

      await expectOutputFile(result.results[0].destination)
    })

    test.skip("renders video from video with cover", async () => {
      const output = outputDefaultPopulate({ outputType: OutputType.Video, cover: true })
      const outputs: CommandOutputs = [output]
      const renderingInput: RenderingInput = renderingInputFromRaw(LoadType.Video, videoUrl)
      const processArgs: RenderingProcessArgs = {
        ...renderingProcessArgs('video-from-video-with-cover'), outputs, ...renderingInput,
      }
      const renderingProcess = renderingProcessInstance(processArgs)
      const result = await renderingProcess.runPromise()
      const { results } = result
      expect(results.length).toBeGreaterThan(0)
      const [first] = results
      expect(first.error).toBeUndefined()
      await expectOutputFile(first.destination)
    })

    test.skip("renders image sequence from video with cover", async () => {
      const output =outputDefaultPopulate({ ...baseOutput, outputType: OutputType.ImageSequence, cover: true })
      // console.log('output', output)
      const outputs: CommandOutputs = [output]
      const renderingInput: RenderingInput = renderingInputFromRaw(LoadType.Video, videoUrl)
      const processArgs: RenderingProcessArgs = {
        ...renderingProcessArgs('image-sequence-from-video-with-cover'), outputs, ...renderingInput,
      }
      const renderingProcess = renderingProcessInstance(processArgs)
      const result = await renderingProcess.runPromise()
      await expectOutputFile(result.results[0].destination)
    })

    // BUG: test renders as transparent image instead of properly cropped
    test.skip("renders image from image", async () => {
      expect('image output does not currently support negative values for overlay.y').toBeFalsy()

      const output = outputDefaultPopulate({ ...baseOutput, cover: false })
      const outputs: CommandOutputs = [output]
      const renderingInput: RenderingInput = renderingInputFromRaw(LoadType.Image, imageUrl)
      const processArgs: RenderingProcessArgs = {
        ...renderingProcessArgs('image-from-image-square'), outputs, ...renderingInput,
      }
      const process = renderingProcessInstance(processArgs)
      const result = await process.runPromise()
      await expectOutputFile(result.results[0].destination)
    })
  })
})
// ffmpeg -y -loop 1 -i /Users/doug/GitHub/moviemasher.js/dev/shared/image/globe.jpg -filter_complex "color=rate=10:color=#00000000:size=320x240[COLORBACK];[0:v]setpts=expr=PTS-STARTPTS[BLA];[BLA]scale=width=1*in_w*max(320/in_w\,240/in_h):height=1*in_h*max(320/in_w\,240/in_h)[SCALE0];[SCALE0]format=pix_fmts=rgba[FORMAT];[FORMAT]setsar[SETSAR0];[COLORBACK][SETSAR0]overlay=eval=init:x=320-overlay_w:y=240-overlay_h" -an -q:v 1 -frames:v 1 temporary/test/render/image-from-image-square/image.png
//ffmpeg -y -loop 1 -i /Users/doug/GitHub/moviemasher.js/dev/shared/image/globe.jpg -filter_complex "color=rate=10:color=#00000000:size=320x240[COLORBACK];[0:v]scale=width=1*in_w*max(320/in_w\,240/in_h):height=1*in_h*max(320/in_w\,240/in_h)[SCALE0];[SCALE0]setsar[SETSAR0];[COLORBACK][SETSAR0]overlay=x=abs(0.5*(320-overlay_w)):y=abs(0.5*(240-overlay_h))" -frames:v 1 temporary/test/render/image-from-image-square/image.png


// ffmpeg -y -loop 1 -i /Users/doug/GitHub/moviemasher.js/dev/shared/image/globe.jpg -filter_complex "color=rate=10:color=#00000000:size=320x240[COLORBACK];[0:v]setpts=expr=PTS-STARTPTS[BLA];[BLA]scale=width=1*in_w*max(320/in_w\,240/in_h):height=1*in_h*max(320/in_w\,240/in_h)[SCALE0];[SCALE0]setsar[SETSAR0];[COLORBACK][SETSAR0]overlay=x=0.5*(320-overlay_w):y=0.5*(240-overlay_h)" -frames:v 1 temporary/test/render/image-from-image-square/image.png
