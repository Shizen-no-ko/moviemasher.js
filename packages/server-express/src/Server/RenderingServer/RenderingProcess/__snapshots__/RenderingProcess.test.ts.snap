// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RenderingProcess renders video-from-color.json 1`] = `
"SHA256=566d7ff8add67fdb4cebfb8b6ba73450d32b85ad333d6ce3ed1e459a337d1048
"
`;

exports[`RenderingProcess renders video-from-colors-fade.json 1`] = `
"SHA256=b407bec05d718b8dfb36409f213d72cd43a28dc858a878c8dc6e4b3d2f827fb2
"
`;

exports[`RenderingProcess renders video-from-image.json 1`] = `
"SHA256=8a169f737a7476c97d5afe7b034dbe742715d1343ae06908c87e0c659824fa26
"
`;

exports[`RenderingProcess renders video-from-image-pan.json 1`] = `
"SHA256=cf17d76075c487acdd34e78c17aa64c553005349990eefab41cfae98f5da8b94
"
`;

exports[`RenderingProcess renders video-from-text.json 1`] = `
"SHA256=f139af6cd41e8dc00cbdf1656936757b58c30a2cadc03951a4ad686895527221
"
`;

exports[`RenderingProcess renders video-from-text-fade.json 1`] = `
"SHA256=0db64fa6f2961f4b0792fbee16cbbf180986a2350004685c189b4a2bd4fd6c2b
"
`;

exports[`RenderingProcess runPromise renders image from image with cover 1`] = `
"SHA256=56c8c4481c7e7492e2d91e0e1553fcdae0723c7ecccddd690fc709bfb57c2ade
"
`;

exports[`RenderingProcess runPromise renders video from image 1`] = `
"SHA256=295943edc08f17e5a6862d5f9d90b1e4e62f42d8fccc888e937609f26caef933
"
`;
